### ISSUE TITLE
Placeholders localization

### ISSUE THREAD

**Issue by gotev** [gotev/android-upload-service#466]
Placeholders contains non localizable text at the moment, as pointed out by https://github.com/gotev/android-upload-service/issues/463#issuecomment-544989334

Make them localizable and customizable

**Comment by gotev** [comment]
@ChristinaGit used taken inspiration from your PR #464

What do you think about it?

**Comment by ChristinaGit** [comment]
Pretty close to what I wanted to do. That was my second thought. This should cover the most commonly needed cases. I think this implementation is acceptable.

_When I was doing the request, I decided that adding my types would be too complicated. Also, when I see  in the library its own types for time, rate, speed, etc., I expect that it will already have predefined formatting methods, like `java.time.format.DateTimeFormatter`. So such a solution would force me to make formatters for each of the types. And use them instead of a common interface with multiple methods. So I did not make such an implementation._

Even this solution still has drawbacks. You cannot use different formatting for multiple occurrences of a value in a string.

Example:
For string
`"Uploaded [[UPLOADED_FILES]] ([[UPLOADED_FILES]]/[[TOTAL_FILES]])."`
I want to get
`"Uploaded 4 files (4/10)"`.
("files" must be a part of placeholder since we need pluralization)
But it is too difficult and uncommon case and I think it is not necessary to implement it (But this of course can be implemented by adding parameters to the placeholders, like
`"Uploaded [[UPLOADED_FILES%1]] ([UPLOADED_FILES%2]]/[[TOTAL_FILES]])."`).
 
_Perhaps you should add another placeholder for **the remaining files**._

The naming is strange though. I want to suggest changing names like this:
- `MinutesSeconds` to `UploadElapsedTime` (Since this class is not yet used anywhere else)
- 
```kotlin
interface LocalizedDataProvider {
    fun minutesSeconds(minutesSeconds: MinutesSeconds): String
    fun uploadRate(uploadRate: UploadRate): String
    fun percent(percent: Int): String
    fun successfullyUploadedFiles(uploadedFiles: Int): String
    fun totalFiles(totalFiles: Int): String
}
``` 
to
```kotlin
interface LocalizationProvider {
    fun uploadElapsedTime(minutesSeconds: UploadElapsedTime): String
    fun uploadRate(uploadRate: UploadRate): String
    fun uploadProgress(percent: Int): String
    fun uploadedFiles(uploadedFiles: Int): String
    fun totalFiles(totalFiles: Int): String
}
``` 
(`LocalizationProvider` vs `NotificationLocalizationProvider` vs `UploadLocalizationProvider` is a hard choice)

- 
```kotlin
enum class UploadRateUnit {
    bitSecond,
    kiloBitSecond,
    megaBitSecond
}
```
to
```kotlin
enum class UploadRateUnit {
    bitPerSecond,
    kibibitPerSecond,
    mebibitPerSecond
}
```
See [wikipedia](https://en.wikipedia.org/wiki/Data-rate_units#Mebibit_per_second) for more info.

**Comment by gotev** [comment]
@ChristinaGit thank you for the detailed review and suggestions!

Renamed:
- `MinutesSeconds` to `UploadElapsedTime`
- `LocalizedDataProvider` to `LocalizationProvider` and its method names
- enum values in `UploadRateUnit` to be: `bitPerSecond`, `kiloBitPerSecond`, `megaBitPerSecond`

Added:
- `[[REMAINING_FILES]]` placeholder.

To accomodate:

> You cannot use different formatting for multiple occurrences of a value in a string.

and other advaced usages, I've moved the `Placeholders.replace` method to `LocalizationProvider` `processPlaceholders` method.

So, for simple custom usages you can comfortably make your implementation by extending both `LocalizationProvider` or `DefaultLocalizationProvider` based on specific needs. For more advanced usages, it's possible to also override the `processPlaceholders`, which is a new feature that opens even for custom placeholders.

For example:
```kotlin
override fun processPlaceholders(message: String?, uploadInfo: UploadInfo): String {
    val string = super.processPlaceholders(message, uploadInfo)
    // custom placeholders processing
    return customString
}
```
or even:
```kotlin
override fun processPlaceholders(message: String?, uploadInfo: UploadInfo): String {
    // custom placeholders processing
    return customString
}
```

What's left out so far based on your suggestions:
- *I expect that it will already have predefined formatting methods, like java.time.format.DateTimeFormatter*. How would you implement it?

At this point, the name `LocalizationProvider` seems inadequate, because it is processing placeholders. Also, the class `Placeholders` seems weird. Its only purpose now is to enumerate default placeholders for ease of use in the code.

First thing which came to my mind is:
- rename the `localization` package to `placeholders` package
- rename `LocalizationProvider` to `PlaceholderProcessor` and then `DefaultLocalizationProvider` to `DefaultPlaceholderProcessor` (ideas for a better name?)
- refactor `Placeholders` class to an `enum class` with string values and move it in `localization` package

Making those changes is easy, but first I need some feedback on those thoughts 😉

**Comment by gotev** [comment]
With the latest modifications: https://github.com/gotev/android-upload-service/compare/4a762982e6214595b3413920647aebfde348a7d4...7df307b006f03f6b603ea96ae094da1a9984b5d1

**Comment by ChristinaGit** [comment]
You don't understand me about naming upload rate. kilo**B**itPerSecond should be kilo**b**itPerSecond. The unit name (bit) after SI prefixes (kilo, mega) must be in lower case (ex. millisecond, not milliSecond; kilogram, not kiloGram).

`PlaceholdersProcessor` looks very strange. If someone decides not to handle predefined placeholders, they will still have to implement abstract methods for formatting, although the user should no longer depend on them (bad for [ISP](https://en.wikipedia.org/wiki/Interface_segregation_principle)).

I'll show you how I do formatting usually. But I don't think this library needs it (really).

_This is a simplified example without optimization._
```kotlin
interface Formatter<Target> {
    fun format(target: Target): String
}

interface Formattable<Formatter> {
    fun format(formatter: Formatter): String
}

class UploadRate(val rateBitPerSecond: Long) : Formattable<UploadRateFormatter> {
    override fun format(formatter: UploadRateFormatter): String =
        formatter.format(this)
}

sealed class UploadRateFormatter(val locale: Locale) : Formatter<UploadRate> {
    companion object {
        fun bitPerSecond(locale: Locale): UploadRateFormatter = BitPerSecond(locale)

        fun kilobitPerSecond(locale: Locale): UploadRateFormatter = KilobitPerSecond(locale)
    }

    private class BitPerSecond(locale: Locale) : UploadRateFormatter(locale) {
        override fun format(target: UploadRate): String {
            val unitNames = getUnitNames(locale)
            val numberFormat = NumberFormat.getInstance(locale)

            return numberFormat.format(target.rateBitPerSecond) +
                   Characters.nonBreakingSpace + 
                   unitNames.bitPerSecond
        }
    }

    private class KilobitPerSecond(locale: Locale) : UploadRateFormatter(locale) {
        override fun format(target: UploadRate): String {
            val unitNames = getUnitNames(locale)
            val numberFormat = NumberFormat.getInstance(locale)

            return numberFormat.format(target.rateBitPerSecond / 1000) +
                   Characters.nonBreakingSpace +
                   unitNames.kilobitPerSecond
        }
    }

    fun getUnitNames(locale: Locale): UnitNames {
        // Or get from database file.
        return if (locale == Locale.ENGLISH) {
            UnitNames(
                bitPerSecond = "b/s",
                kilobitPerSecond = "kb/s"
            )
        } else {
            // TODO: Add other locales.

            throw UnsupportedOperationException()
        }
    }

    class UnitNames(
        val bitPerSecond: String,
        val kilobitPerSecond: String
    )
}

fun UploadRate.formatBitPerSecond(locale: Locale = Locale.getDefault()) =
    format(UploadRateFormatter.bitPerSecond(locale))

fun UploadRate.formatKilobitPerSecond(locale: Locale = Locale.getDefault()) =
    format(UploadRateFormatter.kilobitPerSecond(locale))
```
And usage
```kotlin
fun usage() {
    val rate = UploadRate(12312)

    val formatted = if (rate.rateBitPerSecond < 1000) {
        rate.formatBitPerSecond()
    } else {
        rate.formatKilobitPerSecond()
    }
}
```

**Comment by gotev** [comment]
@ChristinaGit 

> kiloBitPerSecond should be kilobitPerSecond. The unit name (bit) after SI prefixes (kilo, mega) must be in lower case (ex. millisecond, not milliSecond; kilogram, not kiloGram).

Thanks for the explanation 😉 fixed.

> `PlaceholdersProcessor` looks very strange. If someone decides not to handle predefined placeholders, they will still have to implement abstract methods for formatting, although the user should no longer depend on them (bad for ISP).

Right. `PlaceholdersProcessor` is an interface again, with only the method to process placeholders. `DefaultPlaceholdersProcessor` adds the other ones and implements the interface. 

This way:
- for a complete custom placeholders management, implement `PlaceholdersProcessor`
- to build up something on the default implementation, extend `DefaultPlaceholdersProcessor` overriding only what's actually needed.

Let me know what you think about it 😃 

For formatters, thanks for sharing your ideas 🎉 I'll consider them, but first let's close the other open points.

**Comment by gotev** [comment]
@ChristinaGit you can also try the actual implementation in Alpha 4, which is now released: #468

**Comment by ChristinaGit** [comment]
It looks good.

_`PlaceholdersProcessor` is now so abstract that it can replace messages in general (from `UploadNotificationStatusConfig`). The user can now simply specify their own placeholder and replace the entire message. As a result, we came to a solution that actually replaces strings messages with callbacks. Not directly, but through a global config. I think it's okay._

I noticed you use different naming in different enum classes. `Placeholder` - UpperCamelCase, and `UploadRateUnit` - lowerCamelCase. I think it's a mistake. Kotlin suggests [UpperCamelCase](https://kotlinlang.org/docs/reference/coding-conventions.html#property-names) for enums.

**Comment by gotev** [comment]
@ChristinaGit thanks for the feedback!

Regarding the `UpperCamelCase` you're right and I haven't noticed it. It should definitely follow the convention. I expected `ktlint` to take care of that, but as far as I've seen, it's not enforced. Opened an issue: https://github.com/pinterest/ktlint/issues/635

**Comment by ChristinaGit** [comment]
Default `UploadNotificationConfig` can be useful. This is quite a difficult decision. I feel something is wrong when I see the default mutable configurations, which are also passed [implicitly](https://deviq.com/explicit-dependencies-principle/). If I need the same config for all uploads, then I just make a factory for them with that config (which I've injected through the factory constructor). This is not always convenient, so I often see such solutions in libraries. If the configuration can be changed multiple times, then obviously it must be thread-safe (and maybe immutable). This must be implemented very carefully.

Thread safety also needs to be considered for the `UploadServiceConfig`. It is mutable and it creates many problems and questions at once. It reminds me [RxJavaPlugins](https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/plugins/RxJavaPlugins.java).

I think before implementing something like this, you should revisit initialization and configuration of upload service (or you can just ignore my nonsense). In such cases, I usually do something like this (I can write in more detail if you need)
```kotlin
private data class ConfigData(var pool: ThreadPool) : ImmutableBuilder {
   companion object {
      val start = ConfigData(default values)
   }

   fun threadPool(pool: ThreadPool) = copy(pool = pool)
}

object UploadServiceConfig {
    @Volatile
    private var _configData: ConfigData?

    fun initialize(initializer: (ImmutableBuilder) -> ImmutableBuilder) = syncronized {
        if (_configData != null) {
            throw IllegalStateException()
        }

        _configData = initializer(ImmutableBuilder.start)
    }
}


UploadServiceConfig.initialize { immutableBuilder ->
    immutableBuilder
        .threadPool(...)
        .notificationConfig(...)
}
```

 It might be worthwhile to prohibit reinitialization. To prevent situations like this (because programmers are lazy and stupid):
```kotlin
UploadServiceConfig.setDefaultNotificationConfig(config1)

UploadRequest
    .initialize(...)
    .start()

UploadServiceConfig.setDefaultNotificationConfig(config2)
```

I also think that since we have an Android library, I think it is worth mentioning in the documentation that the configuration should not capture the context of the activity, service or widget (or use `WeakReference`). And should _usually_ be executed in the application class.

**Comment by gotev** [comment]
Thanks @ChristinaGit 

In the meantime after the beta 4 and your response, I played around with different ideas. To centralize upload notification config is crucial to also set the notification channel, which has to be set only once (exactly like the namespace). I agree with your analysis of initialization and immutable state. There needs to be a configuration builder which could be called only once. I'd like a Kotlin type-safe builder to do that job, but that is going to sacrifice Java users (Android is moving to Kotlin though and I am getting rid of all the Java in all the projects too).

The major problem to reach that goal arises with the notification actions. 

In the example project there's a simple cancel operation for the upload task, which was added several years ago as an example and it's not part of the core library. Looking at it after so much time feels overly complicated. Just for a sample action you have to:
- create a custom uploadID, just to pass it in the pending broadcast intent. 
- add an action to the notification config status on which you want the action to be displayed (to display the same for all you have to create a factory method)
- create a pending broadcast intent and add some custom parameters together with the upload ID
- write a broadcast receiver and register it
- intercept the intent and deserialize parameters from intent
- finally do your implementation, which may involve passing the uploadID to be cancelled to Upload Service static methods which operates on the active upload queue (like in the demo app example).

It's how Android SDK works, but since this library aims to make working with uploads simpler and intuitive, and cancelling a task is a commonly wanted feature, I think it should be a lot easier than that, but without sacrificing those who wants to have full control and know what they are doing.

Design has to be done carefully, so I think the better will be to make a separate issue and talk about that, which can be added in a future 4.x release. Many things already changed since the 3.x and the Kotlin conversion and many others will be changed, but one step a time is better.

**Comment by ChristinaGit** [comment]
> There needs to be a configuration builder which could be called only once. I'd like a Kotlin type-safe builder to do that job, but that is going to sacrifice Java users (Android is moving to Kotlin though and I am getting rid of all the Java in all the projects too).

I did such builders in java and C# before the existence of Kotlin. I know that this can all be done without sacrificing java support.

**Comment by gotev** [comment]
This is possible by extending either DefaultPlaceholdersProcessor.kt or PlaceholdersProcessor.kt with own custom implementation.
